be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala

```diff
@@ -459,4 +461,150 @@ object MessageDataProtocol extends DefaultJsonProtocol {
+          laoId.convertTo[Hash],
+          serverAddress,
+          publicKey.convertTo[PublicKey],
+          challenge.convertTo[Message]
+        )
+      case _ => throw new IllegalArgumentException(s"Can't parse json value $json to a FederationInit object")
+    }
+
+    override def write(obj: FederationInit): JsValue = {
+      var jsObjectContent: ListMap[String, JsValue] = ListMap[String, JsValue](
+        PARAM_OBJECT -> JsString(obj._object.toString),
+        PARAM_ACTION -> JsString(obj.action.toString),
+        PARAM_ID -> obj.laoId.toJson,
+        PARAM_SERVER_ADDRESS -> obj.serverAddress.toJson,
+        PARAM_PUBLIC_KEY -> obj.publicKey.toJson,
+        PARAM_CHALLENGE -> obj.challenge.toJson
+      )
+      JsObject(jsObjectContent)
+    }
+  }
+
+  implicit object FederationChallengeRequestFormat extends JsonFormat[FederationChallengeRequest] {
+    final private val PARAM_TIMESTAMP: String = "timestamp"
+
+    override def read(json: JsValue): FederationChallengeRequest = json.asJsObject().getFields(PARAM_TIMESTAMP) match {
+      case Seq(timestamp @ JsNumber(_)) => FederationChallengeRequest(
+          timestamp.convertTo[Timestamp]
+        )
+      case _ => throw new IllegalArgumentException(s"Can't parse json value $json to a FederationChallengeRequest object")
+    }
+
+    override def write(obj: FederationChallengeRequest): JsValue = {
+      var jsObjectContent: ListMap[String, JsValue] = ListMap[String, JsValue](
+        PARAM_OBJECT -> JsString(obj._object.toString),
+        PARAM_ACTION -> JsString(obj.action.toString),
+        PARAM_TIMESTAMP -> obj.timestamp.toJson
+      )
+      JsObject(jsObjectContent)
+    }
+  }
+
+  implicit object FederationResultFormat extends JsonFormat[FederationResult] {
+    final private val PARAM_STATUS: String = "status"
+    final private val PARAM_REASON: String = "reason"
+    final private val PARAM_PUBLIC_KEY: String = "public_key"
+    final private val PARAM_CHALLENGE = "challenge"
+
+    override def read(json: JsValue): FederationResult =
+      val jsObj = json.asJsObject
+      val status = jsObj.fields.get(PARAM_STATUS).collect {
+        case JsString(s) => s
+      }.getOrElse(throw new IllegalArgumentException(s"Can't parse json value $json to a FederationResult object"))
+
+      val reason = jsObj.fields.get(PARAM_REASON).collect {
+        case JsString(r) => r
+      }
+
+      val publicKey = jsObj.fields.get(PARAM_PUBLIC_KEY).collect {
+        case key @ JsString(_) => key.convertTo[PublicKey]
+      }
+
+      val challengeMessage = jsObj.fields.get(PARAM_CHALLENGE).collect {
+        case jsObj: JsObject => jsObj.convertTo[Message]
+      }.getOrElse(throw new IllegalArgumentException(s"Can't parse json value $json to a FederationResult object"))
+
+      (reason, publicKey) match {
+        case (None, Some(key)) => FederationResult(status, key, challengeMessage)
+        case (Some(r), None)   => FederationResult(status, r, challengeMessage)
```

Comment:
I believe the public key is a mandatory parameter, is this case really reachable ?

Reply:
the public key is only mandatory when we have success as a result

Reply:
Yes indeed, I misread the doc, perfect !